### DISCUSSION TITLE
How to decouple Identity from Domain Layer?

### DISCUSSION THREAD

**Discussion by HybridSolutions** [jasontaylordev/CleanArchitecture#415] [Q&A]
Would love to hear some opinions on this subject. It's one of the most present topics in this architecture, but never properly explained. Found a project based on this architecture that uses automapper to "sync" a custom User class to IdentityUser user entity but is this the best approach?

[https://github.com/mmacneil/CleanAspNetCoreWebApi](url)

It would be nice to see how CleanArchitecture deals with:

1. having a custom User class in domain layer with custom properties not dependent of IdentityUser.
2. having the need to this custom user entity table be related to IdentityUser AspNetUser table.
3.  sync both entities upon CRUD operations on users and prevent orphan data in case of db errors.

**Comment by Chris-Mingay** [comment]
Not at all claiming this is the right way to go but...

I've been going back and forward with this a lot myself recently and I came to the conclusion that I should use identity framework for two things only and then keep it as a completely separate entity in the system I am writing:

- User Authentication - Allowing a user to prove they are who they say they are
- User Authorization - Defining roles a specific user has and defining what they can and cannot do in the system (at a very basic level)

I have my own internal user class which is the only part of the system that references an Identity User ID (via a `public string IdentityUserId { get; set; }`). This is a property I set manually when creating my own User object. I make no attempt to formally link the user to an identityuser via Entity Framework. I have a Mediatr query that allows me to load a `User` from the provided `IdentityUserId` and that's it.

I felt comfortable doing things this way because in my mind, Identity Server is a separate thing outside the scope of my application, I wanted to keep it as separate as possible. I effectively don't care about any Identity based data for a user, it's none of my business, I just want to use it to allow a user to prove who they are.

I use Authorization to define some very simple roles which basically set the type of user each entity is and with it which pages in the system they can see, however for my project a lot of the access logic is business driven, so I keep that within the domain and nothing to do with Identity Server.

**Reply by HybridSolutions** [reply]
Thanks for the input! How are you dealing with users that get deleted? How do you ensure that both your user entity and IdentityUser entity gets deleted and where? Are you using a transaction for that?

**Reply by Chris-Mingay** [reply]
To be honest I've not added anything in relation to user deletion as it's not required for the project I am working on. However I imagine I would do it via an application level command. I would delete both the domain level user and do whatever is required to delete an identity server user as well.

Actually thinking about it, I probably wouldn't delete the Identity server user, they just wouldn't be granted access to anything and when/if required they can be reactivated internally. Again it's following the notion that Identity server is just a vehicle allowing a user to prove who they are, it doesn't necessarily grant them access to anything. 

Obviously if a user requested complete deletion I would do something, but this is an internal system more than a public one.

**Reply by ali-gol** [reply]
@Chris-Mingay How users name, surname, username information? Or even the e-mail address required for authentication and also needed for Application. 
Since the identity server has all those information, do you also keep them in your internal User table?  
So, your internal table and the identity server has some duplicated(mutual) fields?

**Reply by HybridSolutions** [reply]
I hava a working solution and it works perfectly since allowed to isolate identity completely from the Domain layer and keep Identity AspNetUsers table untouched. Here's how I did it:

1. Created my Users table with my own custom field types (Address, Age, etc) and the UserName field that exists in the AspNetUsers table. This field will be the only one that is common to AspNetUsers table. The password is only stored in AspNetUsers table but I needed to have it as a class property in Users class as well to be able store it while in use by the application.
2. In Infrastructure layer, created a service called UserService that has a CreateUserAsync(User user) method that is responsible for creating a user in the AspNetUsers table using UserManager and create a user in my Users table. This happens inside a transaction so that if anything fails, the operation is rolled back.

For all the operations that require changes to both AspNetUsers and Users tables, I had to make sure the operations could roll back in case something went wrong, Here's the delete method:

`public async Task<ApplicationResult> DeleteUserAsync(int userId)
    {
        User user = await FindByIdAsync(userId);

        if (user != null)
        {
            var identityUser = await _userManager.FindByNameAsync(user.UserName);

            using (IDbContextTransaction transaction = _context.Database.BeginTransaction())
            {
                var result = await _userManager.DeleteAsync(identityUser);

                if (result.Succeeded)
                {
                    await _repository.DeleteAsync(user);
                    transaction.Commit();
                }
                else
                {
                    transaction.Rollback();
                    return ApplicationResult.Failed();
                }
            }
        }

        return ApplicationResult.Success;
    }`

I try to delete the user from AspNetUsers using UserManager, and if succeeds, I try to delete the user from my Users table. 
Hope this helps you out.

**Reply by ali-gol** [reply]
@HybridSolutions I have done the similar. I have the IdentityUserId in my seperated User class.
But if I do it as you did, I won't be able to create CreateUserCommand or those kind of things. I mean in your solution, you won't be able to use Mediatr for User Creation, because you can not call the Infra from Application layer.

**Reply by HybridSolutions** [reply]
> @HybridSolutions I have done the similar. I have the IdentityUserId in my seperated User class. But if I do it as you did, I won't be able to create CreateUserCommand or those kind of things. I mean in your solution, you won't be able to use Mediatr for User Creation, because you can not call the Infra from Application layer.

Yes, but I gave up going strictly by the book, because I would go insane and there's little support on integrating identity in a Clean Architecture solution. Do you really need to implement Mediatr pattern? Would love to see that working.

**Reply by danielmackay** [reply]
Could you simply have the `user` entities in your `domain` project and encapsulate the `IdentityService` by having the interface in `application` and the implementation in your `infrastructure` (as you've  done)

Then have a `CreateUserCommand` that creates both a business user and identity user.

Would that work for your situation?

**Reply by ali-gol** [reply]
> Could you simply have the `user` entities in your `domain` project and encapsulate the `IdentityService` by having the interface in `application` and the implementation in your `infrastructure` (as you've done)
> 
> Then have a `CreateUserCommand` that creates both a business user and identity user.
> 
> Would that work for your situation?

Yes @danielmackay, that would solve the issue. But creating user in identity service is not the responsibility of Application Layer. If it is, what is the benefit of having the Seperate identity server? Simply disable it. 

That's the problem of this solution. It forces you to use the identity service, since we don't need it, we just workaround our issues. 
That's why I've totally disabled the Identity Server. I am basically returning my custom JWT from a login endpoint. This template solution ignores the fact that not everyone needs an identity server. For most of scenarios, it is over engineering.

On the other hand, I don't see any documentation about how we should structure out solution based on real life app. In the given examples and sample todo app; there is no user entity involved. It would be more realistic to have such a thing like that as we discussed here.
Thanks for your reply, and effort.

**Reply by danielmackay** [reply]
@ali-gol - I think the point of using IdentityServer is showing one way to do authentication.  There are also many other options such as ASP.NET Identity, or an OIDC provider such as Azure AD, OAuth, or Okta.  It's hard to build a template that works exactly right for everyone, and sometimes it will get you 90% of the way there, but you may need to remove some pieces and replace with your own (which it sounds like you've done).  That is a very normal thing to do.

I do hear what you're saying about IdentityServer, and I believe @jasontaylordev has been thinking about how to add a more simplified Identity Service that devs can extend in a way that suits different needs.

If you have some ideas feel free to submit your own PR.
